src/main/java/bio/terra/service/load/LoadDao.java

```
@@ -49,72 +44,101 @@ public LoadDao(NamedParameterJdbcTemplate jdbcTemplate, ConfigurationService con
+        return load;
     }
 
-    // To support idempotent steps, this method will not complain if the load no longer exists or if
-    // it is already unlocked. It throws if the lock is held by a different flight or if the unlock operation
-    // fails for some reason.
+    /**
+     * Unlocking means deleting the specific row for our load tag and flight id.
+     * If no row is deleted, we assume this is a redo of the unlock and is OK.
+     * If a row is deleted, then we have performed the unlock. So we don't
+     * check the affected row count.
+     *
+     * @param loadTag tag identifying this load
+     * @param flightId flight id releasing the lock
+     */
     @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
     public void unlockLoad(String loadTag, String flightId) {
-        Load load = lookupLoadByTag(loadTag);
-        if (load == null || !load.isLocked()) {
-            return; // nothing to unlock
-        }
+        jdbcTemplate.getJdbcTemplate().execute("LOCK TABLE load IN EXCLUSIVE MODE");
 
-        if (!StringUtils.equals(load.getLockingFlightId(), flightId)) {
-            conflictThrow(load);
-        }
-        updateLoad(load.getId(), false, null);
-        load = lookupLoadByTag(loadTag);
-        if (load == null || load.isLocked() || load.getLockingFlightId() != null) {
-            throw new LoadLockFailureException("Internal error: failed to unlock a load!");
+        String delete = "DELETE FROM load WHERE load_tag = :load_tag AND locking_flight_id = :flight_id";
+        MapSqlParameterSource params = new MapSqlParameterSource()
+            .addValue("load_tag", loadTag)
+            .addValue("flight_id", flightId);
+        int rowCount = jdbcTemplate.update(delete, params);
+        if (rowCount != 1) {
+            throw new LoadLockedException("Flight does not hold the load lock for load tag: " + loadTag);
         }
```

Regarding the method description, if this is a repeated unlock and no row is deleted, then won't this throw an exception instead of returning OK?

---

Yes, thanks for catching that!